### PR TITLE
storage/engine: Use pebble.Batch to build/iterate over RocksDB batches

### LINKED ISSUE #39672: storage/engine: use pebble.{Batch,BatchReader} for raw batch manipulation
`RocksDBBatchBuilder` and `RocksDBBatchReader` have intimate knowledge of the RocksDB write batch format. So do `pebble.{Batch,BatchReader}`. In fact, they share a lineage. We should replace the raw batch manipulation in `RocksDBBatch{Reader,Builder}` with usage of `pebble.{Batch,BatchReader}` in order to reduce code duplication.

### PR DESCRIPTION
**Pull request by itsbilal**
We currently manipulate with the RocksDB batch format directly in
Cockroach through RocksDBBatch{Reader, Builder}. pebble.Batch follows
the same binary format, so leverage that code instead of duplicating
it in CockroachDB.

Fixes #39672

Depends on changes in https://github.com/petermattis/pebble/pull/231 .

Release note: None

### CODE REVIEW

**Review by petermattis — COMMENTED** [review]
Directionally looks good. My two big comments are about the `finished` boolean and the extra copying caused by encoding keys into a temporary buffer. Adjusting the `pebble.Batch` implementation and interface could remove those contortions. Thoughts?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/39756#-:-LmgD_s3D8OhApcbkEq-:bhmk35y)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajkr, @itsbilal, and @petermattis)

___
*[pkg/storage/engine/batch.go, line 105 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39756#-Lmg9i7M56U3yVqbRlpP:-Lmg9i7M56U3yVqbRlpQ:b-o6vufr) ([raw file](https://github.com/cockroachdb/cockroach/blob/98cb16b0c48319b205101ed41207efc31bf92602/pkg/storage/engine/batch.go#L105)):*
> ```Go
> // is only valid until the next builder method is called.
> func (b *RocksDBBatchBuilder) Finish() []byte {
> 	b.batch.MaybeInit(initialBatchSize)
> ```

I think this call wouldn't be necessary if `Repr` returned an empty batch. This could possibly be a global constant since I don't think there are any guarantees about `Repr()` needing to return memory that is safe for the caller to manipulate.
___
*[pkg/storage/engine/batch.go, line 109 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39756#-LmgADgf-oCfrMUXqmCm:-LmgADgf-oCfrMUXqmCn:bi9f66i) ([raw file](https://github.com/cockroachdb/cockroach/blob/98cb16b0c48319b205101ed41207efc31bf92602/pkg/storage/engine/batch.go#L109)):*
> ```Go
> 	repr := b.batch.Repr()
> 	// Set finished to true, so the next batch operation calls reset().
> 	b.finished = true
> ```

You're not calling `Reset` here because that would mutate `repr`, right? That might argue for changing how `pebble.Batch` tracks count (i.e. pull it out into a member variable and only set it when `Repr()` is called).
___
*[pkg/storage/engine/batch.go, line 154 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39756#-LmgAmgw--hZQxDLDuo1:-LmgAmgw--hZQxDLDuo2:b-slun6d) ([raw file](https://github.com/cockroachdb/cockroach/blob/98cb16b0c48319b205101ed41207efc31bf92602/pkg/storage/engine/batch.go#L154)):*
> ```Go
> 		b.reset()
> 	}
> 	b.buf = EncodeKeyToBuf(b.buf[:0], key)
> ```

Encoding into a temporary buffer and then copying is an unfortunate aspect of this change. I wonder if there are any performance implications. One way to avoid this would be to "decompose" the `Batch.{Set,Delete,Merge,...}` operations. Those operations would still remain, but there would be a parallel set of `Batch.{SetX,DeleteX,MergeX}` operations which would take a key/value len instead of a key/value. These operations would return one or two slices where the key (and value) should be copied (or encoded in this case).
___
*[pkg/storage/engine/batch.go, line 223 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39756#-LmgD0cUBGaX0N42IYyk:-LmgD0cUBGaX0N42IYyl:b-n1ljyt) ([raw file](https://github.com/cockroachdb/cockroach/blob/98cb16b0c48319b205101ed41207efc31bf92602/pkg/storage/engine/batch.go#L223)):*
> ```Go
> 
> 	b2 := pebble.NewBatchFromRepr(repr)
> 	reprCount, _ := RocksDBBatchCount(repr)
> ```

This makes me think a `pebble.Batch.ApplyRepr` method should be added.

**Review by itsbilal — COMMENTED** [review]
I've introduced new methods to `pebble.Batch` to prevent the double copy. We will probably see more instances where we would like to be able to pass `MVCCKey`s to Pebble, so having an `Encodeable` interface seems like a good general solution. PTAL!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/39756#-:-LmkEfAzBdjr5qjNl7Ro:b4xzv94)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajkr and @petermattis)

___
*[pkg/storage/engine/batch.go, line 105 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39756#-Lmg9i7M56U3yVqbRlpP:-Lmjgptl5t--aQV7g4Lt:b20mj2l) ([raw file](https://github.com/cockroachdb/cockroach/blob/98cb16b0c48319b205101ed41207efc31bf92602/pkg/storage/engine/batch.go#L105)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

I think this call wouldn't be necessary if `Repr` returned an empty batch. This could possibly be a global constant since I don't think there are any guarantees about `Repr()` needing to return memory that is safe for the caller to manipulate.
</blockquote></details>

Given the batch sizes are the same, I'll take this call away and roll the initialization check into `Repr`.
___
*[pkg/storage/engine/batch.go, line 109 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39756#-LmgADgf-oCfrMUXqmCm:-Lmjbbay5NvIBC8wdlcW:b3rd85u) ([raw file](https://github.com/cockroachdb/cockroach/blob/98cb16b0c48319b205101ed41207efc31bf92602/pkg/storage/engine/batch.go#L109)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

You're not calling `Reset` here because that would mutate `repr`, right? That might argue for changing how `pebble.Batch` tracks count (i.e. pull it out into a member variable and only set it when `Repr()` is called).
</blockquote></details>

Yes, that's why. I actually like the way how we don't duplicate `count` in a lot of places inside `pebble.Batch` - writing back the `count` to the repr ever time we called `Repr` was a bit trippy in the old batch builder. But then again, we go ahead and duplicate `count` here because `rocksdb.go` depends on it.

What do you think? I'd be in favour of keeping the finished boolean as it lets us keep a single source of truth for count in `pebble.Batch`.
___
*[pkg/storage/engine/batch.go, line 154 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39756#-LmgAmgw--hZQxDLDuo1:-LmjimKf02sw5B_3tQed:be502pw) ([raw file](https://github.com/cockroachdb/cockroach/blob/98cb16b0c48319b205101ed41207efc31bf92602/pkg/storage/engine/batch.go#L154)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Encoding into a temporary buffer and then copying is an unfortunate aspect of this change. I wonder if there are any performance implications. One way to avoid this would be to "decompose" the `Batch.{Set,Delete,Merge,...}` operations. Those operations would still remain, but there would be a parallel set of `Batch.{SetX,DeleteX,MergeX}` operations which would take a key/value len instead of a key/value. These operations would return one or two slices where the key (and value) should be copied (or encoded in this case).
</blockquote></details>

Done. Good point - the double copy should ideally be avoided.
___
*[pkg/storage/engine/batch.go, line 223 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39756#-LmgD0cUBGaX0N42IYyk:-LmjiRlD95_YQzZ2LLnf:b-n141ke) ([raw file](https://github.com/cockroachdb/cockroach/blob/98cb16b0c48319b205101ed41207efc31bf92602/pkg/storage/engine/batch.go#L223)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

This makes me think a `pebble.Batch.ApplyRepr` method should be added.
</blockquote></details>

I like the fact that `Apply` takes a `Batch`, and that's existing `pebble.Batch` code that's already in place. Think it makes more sense to use that and `SetRepr` instead of a specialized `ApplyRepr` that will only be used here.

Updated this to use `SetRepr`

**Review by petermattis — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/39756)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajkr, @itsbilal, and @petermattis)

___
*[pkg/storage/engine/batch.go, line 109 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39756#-LmgADgf-oCfrMUXqmCm:-LmklifP5MQouT5kk57a:b3jln8g) ([raw file](https://github.com/cockroachdb/cockroach/blob/98cb16b0c48319b205101ed41207efc31bf92602/pkg/storage/engine/batch.go#L109)):*
<details><summary><i>Previously, itsbilal (Bilal Akhtar) wrote…</i></summary><blockquote>

Yes, that's why. I actually like the way how we don't duplicate `count` in a lot of places inside `pebble.Batch` - writing back the `count` to the repr ever time we called `Repr` was a bit trippy in the old batch builder. But then again, we go ahead and duplicate `count` here because `rocksdb.go` depends on it.

What do you think? I'd be in favour of keeping the finished boolean as it lets us keep a single source of truth for count in `pebble.Batch`.
</blockquote></details>

There seems to be a wart here in that `RocksDBBatchBuilder.count` still exists. I think the single source of truth for the number of elements in a batch should be in `pebble.Batch`. How does `rocksdb.go` depend on the `count` field. Is it just accessing it? Couldn't we replace those accesses with a `RocksDBBatchBuilder.Count()` method?
___
*[pkg/storage/engine/batch.go, line 154 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39756#-LmgAmgw--hZQxDLDuo1:-Lmklayn779MWpCcM4Mf:b-uw2abf) ([raw file](https://github.com/cockroachdb/cockroach/blob/98cb16b0c48319b205101ed41207efc31bf92602/pkg/storage/engine/batch.go#L154)):*
<details><summary><i>Previously, itsbilal (Bilal Akhtar) wrote…</i></summary><blockquote>

Done. Good point - the double copy should ideally be avoided.
</blockquote></details>

LGTM modulo my comments in the Pebble PR about the potential overhead of this interface.

**Review by itsbilal — COMMENTED** [review]
Rebased on the pebble PR, then some more minor changes to make tests pass and the linter happy. PTAL!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/39756#-:-LmuLfx04I_yKw7YTa7F:b-38l7ni)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajkr and @petermattis)

___
*[pkg/storage/engine/batch.go, line 109 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39756#-LmgADgf-oCfrMUXqmCm:-LmuL5Rm3shl6SHskBMj:b-nxx8d9) ([raw file](https://github.com/cockroachdb/cockroach/blob/98cb16b0c48319b205101ed41207efc31bf92602/pkg/storage/engine/batch.go#L109)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

There seems to be a wart here in that `RocksDBBatchBuilder.count` still exists. I think the single source of truth for the number of elements in a batch should be in `pebble.Batch`. How does `rocksdb.go` depend on the `count` field. Is it just accessing it? Couldn't we replace those accesses with a `RocksDBBatchBuilder.Count()` method?
</blockquote></details>

Done - yes, it was just accessing it. There was one quirk in the previous implementation though, that we rely on pretty heavily - `count` gets zeroed after a Finish() even if the underlying repr has a nonzero count. Our current approach of reusing Reprs to prevent copying meant that we could no longer rely on the correct count coming from `pebble.Batch`, even after your recent PR. Easiest and least intrusive fix is to just have a special case in `Count()` that returns 0 when `finished`.

**Review by petermattis — APPROVED** [review]
:lgtm: modulo my comment about the `finished` field. I think we can get rid of that field once my Pebble PR is merged.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/39756#-:-LmuTO4J-xlVUEcf5h-8:bt7yyiv)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @ajkr and @itsbilal)

___
*[pkg/storage/engine/batch.go, line 109 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39756#-LmgADgf-oCfrMUXqmCm:-LmuSWGZ6uTEyetymlyC:b-amrtxx) ([raw file](https://github.com/cockroachdb/cockroach/blob/98cb16b0c48319b205101ed41207efc31bf92602/pkg/storage/engine/batch.go#L109)):*
<details><summary><i>Previously, itsbilal (Bilal Akhtar) wrote…</i></summary><blockquote>

Done - yes, it was just accessing it. There was one quirk in the previous implementation though, that we rely on pretty heavily - `count` gets zeroed after a Finish() even if the underlying repr has a nonzero count. Our current approach of reusing Reprs to prevent copying meant that we could no longer rely on the correct count coming from `pebble.Batch`, even after your recent PR. Easiest and least intrusive fix is to just have a special case in `Count()` that returns 0 when `finished`.
</blockquote></details>

After github.com/petermattis/pebble#242 goes in, I think you can call `b.batch.Reset()` here and get rid of the `finished` variable.
___
*[pkg/storage/engine/batch.go, line 223 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39756#-LmgD0cUBGaX0N42IYyk:-LmuSp248egiuTpeqJOn:b9jdlup) ([raw file](https://github.com/cockroachdb/cockroach/blob/98cb16b0c48319b205101ed41207efc31bf92602/pkg/storage/engine/batch.go#L223)):*
<details><summary><i>Previously, itsbilal (Bilal Akhtar) wrote…</i></summary><blockquote>

I like the fact that `Apply` takes a `Batch`, and that's existing `pebble.Batch` code that's already in place. Think it makes more sense to use that and `SetRepr` instead of a specialized `ApplyRepr` that will only be used here.

Updated this to use `SetRepr`
</blockquote></details>

The downside to this API is the allocation of a `pebble.Batch`. Given that `Batch.Apply` pretty much just uses `Batch.Repr()`, an API which avoided the allocation could be useful. I don't feel strongly about this needing to be done in this PR, though.

**Review by petermattis — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/39756)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @ajkr and @itsbilal)

___
*[pkg/storage/engine/batch.go, line 89 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/39756#-LmuW46Z7PhFgsbHkXeX:-LmuW46Z7PhFgsbHkXeY:b-1u6vq8) ([raw file](https://github.com/cockroachdb/cockroach/blob/c473fffad4e9b92e4c89b4b8fd5f95cf3041b5a1/pkg/storage/engine/batch.go#L89)):*
> ```Go
> type RocksDBBatchBuilder struct {
> 	batch    pebble.Batch
> 	logData  bool
> ```

For another PR: we should get rid of this `logData` field, and add a `Batch.Empty()` method. The only reason this is present is to ensure we flush a batch which has no records other than `logData`. `Batch.Empty()` would return `len(Batch.storage.data) <= batchHeaderLen`. I'll work on adding that method to Pebble, as there are a few places in Pebble already that want something similar.

**Review by itsbilal — COMMENTED** [review]
Thanks for the review! Seeing the immense amount of special casing we had to do for the `finished` case (especially around counts, a particularly weird discovery) I appreciate the approach of tracking it separately in a variable inside pebble.Batch. Dropped `finished` as you suggested, and updated `vendor` to point to latest pebble.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/39756#-:-Lmucgxv2jAX1eAhgP0z:btxn1xa)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @ajkr, @itsbilal, and @petermattis)

**Review by petermattis — APPROVED** [review]
Much cleaner! :shipit: 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/39756#-:-Lmudo5p71G9sAaaDpmB:bs36q2o)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @ajkr and @itsbilal)

___
*[pkg/storage/engine/rocksdb.go, line 1691 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/39756#-LmuecVb8aR9o4XVAAGr:-LmuecVb8aR9o4XVAAGs:b16cme4) ([raw file](https://github.com/cockroachdb/cockroach/blob/4605e88e4c951edeb1e736628cf1bf06de2c1306/pkg/storage/engine/rocksdb.go#L1691)):*
> ```Go
> 
> 	// Zero all the remaining fields individually. We can't just copy a new
> 	// struct onto r, since r.builder has a sync.NoCopy.
> ```

Hmm, this is unfortunate. I wonder if there is a way to work around it. Just thinking out loud. Don't let this hold up this PR.

**Comment by itsbilal** [comment]
TFTR!

bors r+

**Comment by itsbilal** [comment]
Build failure is https://github.com/cockroachdb/cockroach/issues/39816 and unrelated to this PR.

bors r+

**Pull request merged**
